### PR TITLE
Expanded plugin constructor capabilities

### PR DESCRIPTION
**Pull request by dlvenable**
### Description

* Introduces the `@DataPrepperPluginConstructor` annotation
* When creating a plugin instance, use the `@DataPrepperPluginConstructor` annotated-constructor. If not, then use the constructor which takes the `PluginSetting` or configured model class. Finally, use the default constructor as the last option.
* The plugin framework can also inject a `PluginMetrics` object in the constructor.
* Updated some of the simple plugins to use the default constructor where appropriate.
* Updated `HTTPSource` to use a configuration model. It is provided into the constructor, and the `PluginMetrics` is also now injected.
 
### Issues Resolved

This has some of the proposed functionality from #469.
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginCreator.java
Comment (cmanning09) [inline data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginCreator.java:70]: What happens if there are two separate single argument constructors? What is the failure mode?

What is your vision for v2.0? Do you plan on supporting all of these scenarios in the future as well? Would it be possible to move to an explicit model which requires an annotated constructor?

File: data-prepper-plugins/common/src/main/java/com/amazon/dataprepper/plugins/PluginFactory.java
Comment (cmanning09) [reply data-prepper-plugins/common/src/main/java/com/amazon/dataprepper/plugins/PluginFactory.java:42]: I am fine with this as well because of the reasons you provided. Thanks!

**Review by cmanning09 — CHANGES_REQUESTED** [review]

File: data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginArgumentsContext.java
Comment (dlvenable) [reply data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginArgumentsContext.java:60]: It is the custom configuration supplied which is the type directed by this field: https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-api/src/main/java/com/amazon/dataprepper/model/annotations/DataPrepperPlugin.java#L80

File: data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginCreator.java
Comment (dlvenable) [reply data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginCreator.java:21]: I went through a few class names and it appears that IntelliJ didn't pick up renaming the parameter here. I'll fix this up to match.

File: data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginArgumentsContext.java
Comment (dlvenable) [reply data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginArgumentsContext.java:29]: Yes, the `singleParameterObject` is going to be the custom POJO type or the `PluginSetting`.

File: data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginArgumentsContext.java
Comment (dlvenable) [reply data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginArgumentsContext.java:65]: Yes, I can add this next.

File: data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginArgumentsContext.java
Comment (dlvenable) [reply data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginArgumentsContext.java:34]: To answer your second question, yes, we are support that type of constructor. But, only if it is annotated with `@DataPrepperConstructor`. The `HTTPSource` shows something very similar to this.

I'm not quite sure what your first question is asking. At this point of the code we are only adding a `Supplier`. So, no `PluginMetric` is constructor unless the annotated constructor requests it.

File: data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginCreator.java
Comment (dlvenable) [reply data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginCreator.java:70]: > What happens if there are two separate single argument constructors?
Because the `@DataPrepperPlugin` allows configuring the `pluginConfigurationType`, the single argument constructor will only match that type. So it should only ever be what is defined by `pluginConfigurationType`.

> What is your vision for v2.0? Do you plan on supporting all of these scenarios in the future as well? Would it be possible to move to an explicit model which requires an annotated constructor?
For 2.0, I'd like to remove the single argument constructor. Thus, the constructor must be either annotated with `@DataPrepperConstructor` or be the default constructor.
We need to keep at least support for a single argument `PluginSetting` in 1.x since that is the current behavior. I could change this so that the only implicit single argument constructor is `PluginSetting` and disallow the POJO-based single-argument constructor. Would that help?

Comment (dlvenable) [reply data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginCreator.java:70]: @cmanning09 ,

I'm going to change this behavior so that the only single parameter constructor supported is `PluginSetting`. Thus, plugins which use a POJO for the configuration must provide the `@DataPrepperConstructor` annotation. I believe this should be simpler and is one fewer behavior to deprecate.

File: data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSource.java
Comment (chenqi0805) [inline data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSource.java:35]: QUES: 

(1) What will happen if the pluginType does not match the implemented interface?
(2) Looks like this introduces two sources of truth for pluginType. What is the long term plan?

File: data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSourceConfig.java
Comment (chenqi0805) [inline data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSourceConfig.java:36]: Nit: can we use REQUEST_TIMEOUT constant here?

File: data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSourceConfig.java
Comment (chenqi0805) [inline data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSourceConfig.java:25]: Why can we remove this?

File: data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSourceConfig.java
Comment (chenqi0805) [inline data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSourceConfig.java:39]: Why no jsonProperty annotation on top of this variable?

File: data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSourceConfig.java
Comment (chenqi0805) [inline data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSourceConfig.java:40]: Same as above

File: data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSourceConfig.java
Comment (dlvenable) [reply data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSourceConfig.java:39]: The default conversion will support snake case, so this should serialize. But, I can add add it to be somewhat more thorough and consistent.

File: data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSource.java
Comment (dlvenable) [reply data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSource.java:35]: If `pluginType` does not match the implemented interface, then the plugin framework will throw an exception at runtime. I added the `pluginType` in another PR a few weeks ago. I opted for the approach of having the `pluginType` and the interface because I want plugins to be able to implement other interfaces if they find appropriate.

Comment (chenqi0805) [reply data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSource.java:35]: Thanks. I missed that part.

> I opted for the approach of having the pluginType and the interface because I want plugins to be able to implement other interfaces if they find appropriate.

Could you provide a toy example for this use case? b/c I am still confused how one can avoid `implements Source<Record<String>>` since it involves the data type claim.

Comment (chenqi0805) [reply data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSource.java:35]: Oh. Wait. My mis. Java supports multiple interface but not multiple inheritance.

File: data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSourceConfig.java
Comment (dlvenable) [reply data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSourceConfig.java:25]: This string is no longer being used because the tests use a POJO. I pushed a change which added the string literal into the `@JsonProperty`. For those properties which are being used elsewhere, I kept those as constants.

File: data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginArgumentsContext.java
Comment (dlvenable) [reply data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginArgumentsContext.java:59]: I have added the documentation and linked it from the Development guide. As plugin support is extended, I would like to reconsidering the linking, but I believe this is sufficient for now.

File: data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginCreator.java
Comment (dlvenable) [reply data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginCreator.java]: I pushed a commit which, among other changes, now only permits a single-parameter constructor for `PluginSetting`. To use the custom POJO configuration, plugin authors must add the `@DataPrepperPluginConstructor` annotation.

File: data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSourceConfig.java
Comment (cmanning09) [inline data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSourceConfig.java:53]: Is there a difference between upper case and lower case JsonProperties? I am wondering why `ssl_key_password` and `SSL_KEY_FILE` are different.

Comment (dlvenable) [reply data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSourceConfig.java:53]: `SSL_KEY_FILE` is a static final field whereas `"ssl_key_password"` is a string literal. The former is used in some error strings so I kept it as constant. We could make them all constants if we wanted.

**Comment by dlvenable** [comment]
I also had left some commented out code in place. I just pushed a commit which cleaned that up.

File: data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSourceConfig.java
Comment (cmanning09) [reply data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSourceConfig.java:53]: oh, I missed the quotes. For a second I thought the properties were defined in a different format. I'd prefer consistency but will defer to your preference here. I don't want to hold this up.

**Review by cmanning09 — APPROVED** [review]

**Review by chenqi0805 — APPROVED** [review]

**Pull request merged**
